agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentWebSocketClient.kt

```diff
@@ -0,0 +1,211 @@
+package com.luxoft.blockchainlab.corda.hyperledger.indy
+
+import com.fasterxml.jackson.databind.node.ObjectNode
+import com.luxoft.blockchainlab.hyperledger.indy.utils.SerializationUtils
+import mu.KotlinLogging
+import org.java_websocket.client.WebSocketClient
+import org.java_websocket.handshake.ServerHandshake
+import rx.*
+import java.net.URI
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.ConcurrentLinkedDeque
+
+class AgentWebSocketClient(serverUri: URI, private val socketName: String) : WebSocketClient(serverUri) {
+    private val log = KotlinLogging.logger {}
+
+    override fun onOpen(handshakedata: ServerHandshake?) {
+        log.info { "$socketName:AgentConnection opened: $handshakedata" }
+    }
+
+    override fun onClose(code: Int, reason: String?, remote: Boolean) {
+        log.info { "$socketName:AgentConnection closed: $code,$reason,$remote" }
+    }
+
+    /**
+     * inbound messages list
+     */
+    private val receivedMessages = ConcurrentHashMap<String, ConcurrentLinkedDeque<String>>()
+
+    /**
+     * message routing map
+     */
+    private val subscribedObservers = ConcurrentHashMap<String, ConcurrentLinkedDeque<SingleSubscriber<in String>>>()
+
+    /**
+     * Adds an observer to (FIFO) queue corresponding to the given key.
+     * If the queue doesn't exist for the given key, it's been created.
+     */
+    private fun addObserver(key: String, observer: SingleSubscriber<in String>) =
+            subscribedObservers.getOrPut(key) { ConcurrentLinkedDeque() }.addLast(observer)
+
+    /**
+     * Adds a message to the queue corresponding to the given key.
+     * If the queue doesn't exist for the given key, it's been created.
+     */
+    private fun storeMessage(key: String, message: String) =
+            receivedMessages.getOrPut(key) { ConcurrentLinkedDeque() }.addLast(message)
+
+    /**
+     * Removes an observer from the queue.
+     */
+    private fun popObserver(key: String) =
+            subscribedObservers.getOrPut(key) { ConcurrentLinkedDeque() }.pollFirst()
+
+    /**
+     * Pops a message by key from the queue.
+     */
+    private fun popMessage(key: String) =
+            receivedMessages.getOrPut(key) { ConcurrentLinkedDeque() }.pollFirst()
+
+    /**
+     * Dispatches the message to an observer.
+     * Each message '@type' has own routing agreement
+     */
+    override fun onMessage(msg: String?) {
+        var message = msg
+        log.info { "$socketName:ReceivedMessage: $message" }
+        if (message != null) {
+            val obj = SerializationUtils.jSONToAny<ObjectNode>(message)
+            val type: String = obj["@type"].asText()
+            var key: String? = null
+            when (type) {
```

Comment:
We store by type everything except for those (4 message types) we explicitly specify the storage key for. We don't handle certain responses such as MESSAGE_SENT and RESPONSE_SENT as agreed with @adoroganov. This is to be done in the future versions.
Message specific "key" is either "type" or "type.pubkey", "type.DID", "class.DID" etc. It can't be null, so there's no use case for throwing an exception.